tokenclient/tokenFlows.go

```diff
@@ -0,0 +1,177 @@
+	clientIDParameter          string = "client_id"
+	clientSecretParameter      string = "client_secret"
+)
+
+// NewTokenFlows initializes token flows
+//
+// identity provides credentials and url to authenticate client with identity service
+// options specifies rest client and its tls config, both can be overwritten.
+// Note: Setup of default tls config is not supported for windows os. Module crypto/x509 supports SystemCertPool with go 1.18 (https://go-review.googlesource.com/c/go/+/353589/)
+func NewTokenFlows(identity env.Identity, options Options) (*TokenFlows, error) {
+	t := new(TokenFlows)
+	t.identity = identity
+	if options.HTTPClient == nil {
+		if options.TLSConfig == nil && identity.IsCertificateBased() {
+			defaultConfig, err := defaultTLSConfig(identity)
+			if err != nil {
+				return nil, err
+			}
+			options.TLSConfig = defaultConfig
+		}
+		options.HTTPClient = defaultHTTPClient(options.TLSConfig)
+	}
+	t.options = options
+	t.tokenURI = identity.GetURL() + tokenEndpoint
+	return t, nil
+}
+
+// ClientCredentials implements the client credentials flow (RFC 6749, section 4.4).
+// Clients obtain an access token outside of the context of a user.
+// It is used for non interactive applications (a CLI, a batch job, or for service-2-service communication) where the token is issued to the application itself,
+// instead of an end user for accessing resources without principal propagation.
+//
+// ctx carries the request context like the deadline or other values that should be shared across API boundaries. Default: context.TODO is used
+// customerTenantURL like "https://custom.accounts400.ondemand.com" gives the host of the customers ias tenant
+// options allows to provide a request context and optionally additional request parameters
+func (t *TokenFlows) ClientCredentials(ctx context.Context, customerTenantURL string, options RequestOptions) (string, error) {
+	data := url.Values{}
+	data.Set(grantTypeParameter, grantTypeClientCredentials)
+	data.Set(clientIDParameter, t.identity.GetClientID())
+	if t.identity.GetClientSecret() != "" {
+		data.Set(clientSecretParameter, t.identity.GetClientSecret())
+	}
+	for name, value := range options.Params {
+		data.Set(name, value) // potentially overwrites data which was set before
+	}
+	targetURL, err := t.getURL(customerTenantURL)
+	if err != nil {
+		return "", err
+	}
+	r, e := http.NewRequestWithContext(ctx, http.MethodPost, targetURL, strings.NewReader(data.Encode())) // URL-encoded payload
+	if e != nil {
+		return "", fmt.Errorf("error performing client credentials flow: %w", e)
+	}
+	r.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	tokenJSON, err := t.performRequest(r)
+	if err != nil {
+		return "", err
+	}
+	var response tokenResponse
+	_ = json.Unmarshal(tokenJSON, &response)
+	if response.Token == "" {
+		return "", fmt.Errorf("error parsing requested client credential token: %v", string(tokenJSON))
```

Comment:
This could potentially print a token into the logs.

Reply:
mhh.. this is only true in case identity service returns a token in a different json format, e.g. response does not contain "access_token" property.

Reply:
maybe i could print only the first - lets say 200 characters?

Reply:
Mh not quite happy playing with odds in this case. Maybe @jkbschmid has a better idea.

Reply:
In my opinion, just logging a constant error message would be enough, since this is quite the edge case. If you would prefer more explicit error logging, how about changing the log message to `"access_key" not present in token payload`?
Also, for safety, I think checking for `err != nil` would be good, even if there's already a check for json validity in another method.

Reply:
ok. done